Lbann cached cmake

bvanessen opened the pull request:
Updates the CachedCMakePackage to reduce redundant code in inherited packages:
- Add the CMake prefix path to the cache so that dependent packages are found.
- Enhancements to ensure that CachedCMakePackages get boiler plate fields set in the base level package for both ROCm and CUDA package types.  
- Additionally, it provides a option to make the cached options from variants, bringing it in line with the standard CMakePackage.  
- It also changes the default behavior to force the cached options to overwrite an existing CMakeCache.txt file if one existed.

Added an implementation of the CachedCMakePackage for the LBANN package.

bvanessen committed:
Added support for exporting the CMAKE_PREFIX_PATH and PKG_CONFIG_PATH
to the cached CMake file.  Also split the PATH from being colon
separated to semi-colon separated.

bvanessen committed:
Switching the LBANN recipe to being a CachedCMakePackage.

bvanessen committed:
Added convenience function to create cached field from variant.

Removed inclusion of PKG_CONFIG_PATH since it is mostly redundant with
CMAKE_PREFIX_PATH.

bvanessen committed:
Tweaking the LBANN recipe to use the new define_cmake_cache_from_variant function.

bvanessen committed:
Cleaned up the variables set in the cached_cmake.py file for CUDA.

Added support for using the Ninja generator in the Cached CMake
recipe.  Added the path to the python executable and python path to
cached fields.

bvanessen committed:
Changed the order of the hostname and label to sort by hostname first.

bvanessen committed:
Cleaned up the HIP config entries

bvanessen committed:
Switching to using the define_cmake_cache_from_variant

bvanessen committed:
Export the high performance linkers.

Configure lbann_pfe.sh to use python3.

bvanessen committed:
Changed the default behavior of the CachedCMakeFile build package to
overwrite a local CMakeCache file.  This should ensure that the Spack
configuration is actually respected.

Removed dead code.

bvanessen committed:
Fixed the flags for several cmake cache options to use booleans not
strings.

Removed code formerly in the main cmake setup functions.

bvanessen committed:
Set the HIP_CXX_COMPILER and HIP_ROOT_DIR explicitly for CachedCMakeFile packages.

bvanessen committed:
Fixed the guard for ROCm packages using the CachedCMakePackage to test
for the +rocm variant and not check for a dependency on ^rocm package
in the spec (since no rocm package exists).

bvanessen commented:
@spackbot style fix

bvanessen commented:
@spackbot commands

bvanessen commented:
@spackbot fix style

bvanessen committed:
[@spackbot] updating style on behalf of bvanessen

bvanessen committed:
Removed unused field.

white238 commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -195,26 +231,57 @@ def initconfig_hardware_entries(self):
-            cudacompiler = "${CUDA_TOOLKIT_ROOT_DIR}/bin/nvcc"
-            entries.append(cmake_cache_path("CMAKE_CUDA_COMPILER", cudacompiler))
+            entries.append(cmake_cache_path("CUDAToolkit_ROOT", cudatoolkitdir))
+            entries.append(cmake_cache_path("CMAKE_CUDA_COMPILER", f"{cudatoolkitdir}/bin/nvcc"))
```
```suggestion
            entries.append(cmake_cache_path("CMAKE_CUDA_COMPILER", "${CUDAToolkit_ROOT}/bin/nvcc"))
```
I prefer to reuse the variables, but I'm fine with either way.

bvanessen committed:
Update lib/spack/spack/build_systems/cached_cmake.py

Co-authored-by: Chris White <white238@llnl.gov>

bvanessen commented:
@white238 Can you please take a quick look at this again.  I just merged in your suggestion and then rebased it on the head of develop.

white238 reviewed (dismissed):
Looks fine to me. I don't have time to test it atm because the TOSS4 conversion has caused me problems.

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -13,21 +14,24 @@
 
     value = "ON" if boolean_value else "OFF"
-    return 'set({0} {1} CACHE BOOL "{2}")\n'.format(name, value, comment)
+    force_str = "FORCE" if force else ""
```
A few thoughts:

It looks like adding "force" is a new behavior: it is not ever added by any build system (although I do see a few packages in builtin doing this); it is also set as the default behavior (since `force=True` is the default). According to the docs, this will set the value even if it has been set before. I think this means that you cannot override cache entries by setting cmake options (since options provided to the invocation of `cmake` would come first).

Is there a reason you are introducing this behavior as a default?

bvanessen replied:
@scheibelp So the behavior that I experienced is that if there is an existing `CMakeCache.txt` file then without the force, the config file's behavior is ignored.  I will check to see how command line arguments will interact.

scheibelp replied:
AFAIK `CMakeCache.txt` is typically not distributed with the source, I wanted to check if that's also true for `lbann` (or any other packages you are building).

(this could still be an issue in Spack if you restart after a failed install and changing the package.py to generate a different `.cmake` file, but I assume that would be rare)

bvanessen replied:
@scheibelp Here is a response from Tom. 
```
If you do
cmake -DSOME_VAR=ON_or_OFF -C my_cache.cmake ...
then the FORCE will cause SOME_VAR to have the value in my_cache.cmake. However,
cmake -C my_cache.cmake -DSOME_VAR=ON_or_OFF ...
will cause SOME_VAR to have the value specified on the command line.
Do note that this was experimentally determined using CMake 3.22.1 with all CMake policies set to match that version. 
```

bvanessen replied:
I will change the default so that it does not set the `FORCE` flag.

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -63,6 +67,38 @@ def cache_name(self):
+            raise KeyError('"{0}" is not a variant of "{1}"'.format(variant, self.pkg.name))
+
+        if variant not in self.pkg.spec.variants:
+            return ""
```
(question) Any variant in `self.pkg.variants` should be set in `spec.variants` (since this is concrete): did you find you get an error if you don't handle this case specially?

bvanessen replied:
I modeled this code after the existing function in the `cmake` package.

scheibelp replied:
Thanks for the pointer! This is necessary for conditional variants (as explained in https://github.com/spack/spack/commit/c5aee4d9b4840bc23041fa5d3339012598619cca for the corresponding line in `cmake.py`).

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -195,26 +231,57 @@ def initconfig_hardware_entries(self):
+                    cmake_cache_string("CMAKE_CUDA_ARCHITECTURES", "{0}".format(arch_str))
+                )
+
+        # Since there is no rocm package, look for the variant
```
Note that the equivalent to `if spec.satisfies("^cuda"):` would be `if spec.satisfies("^hip"):`, but testing for `+rocm` is better: if you have `X~rocm -> Y+rocm`, we wouldn't want to set ROCm-related CMake variables for X (I'd argue the same for `+cuda` vs. `^cuda`, but that predates this PR so I won't request changing that).

Therefore I think this comment should be removed.

bvanessen replied:
Okay, I will remove the comment.

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -195,26 +231,57 @@ def initconfig_hardware_entries(self):
-            entries.append(cmake_cache_path("CUDA_TOOLKIT_ROOT_DIR", cudatoolkitdir))
-            cudacompiler = "${CUDA_TOOLKIT_ROOT_DIR}/bin/nvcc"
-            entries.append(cmake_cache_path("CMAKE_CUDA_COMPILER", cudacompiler))
+            entries.append(cmake_cache_path("CUDAToolkit_ROOT", cudatoolkitdir))
```
According to [1](https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html), [2](https://cmake.org/cmake/help/latest/module/FindCUDA.html):

* `CUDA_TOOLKIT_ROOT_DIR` was part of `find_cuda`, which was deprecated for `cmake@3.10:`
* `CUDAToolkit_ROOT` is supported starting with `cmake@3.17:`

You may want to make a note of this (and possibly add a conflict like `conflicts(cmake@:3.16)`, although that's not strictly necessary since the variable would just be ignored for earlier CMake versions.

white238 replied:
Currently removing `CUDA_TOOLKIT_ROOT_DIR` would break any BLT packages since it partially still relies on the deprecated behavior. There is a PR to fix this but it isn't fully tested.

We could output it in those packages though.

white238 replied:
The safest thing is to output both. Deprecated doesn't mean people don't use it. 

But I can work around whatever is decided.

bvanessen replied:
So my packages don't need the deprecated variable, but I left it in for the BLT packages.  I would say that we leave it in for now.

scheibelp replied:
To be clear, do you mean you will add it back to this PR? I only see it being removed from this code block and don't see it added back anywhere else (and don't see any other references in `cmake.py` on the develop branch).

bvanessen replied:
@scheibelp @white238 I added it to the PR.

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -195,26 +231,57 @@ def initconfig_hardware_entries(self):
 
     def std_initconfig_entries(self):
+        cmake_prefix_path_env = os.environ["CMAKE_PREFIX_PATH"]
+        cmake_prefix_path_array = cmake_prefix_path_env.split(":")
```
You'll need to split on `os.pathsep` here rather than `":"` (can be different on Windows vs. Unix).

bvanessen replied:
Done.

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -195,26 +231,57 @@ def initconfig_hardware_entries(self):
             "#------------------{0}".format("-" * 60),
             "# CMake executable path: {0}".format(self.pkg.spec["cmake"].command.path),
             "#------------------{0}\n".format("-" * 60),
+            cmake_cache_path("CMAKE_PREFIX_PATH", cmake_prefix_path),
```
Note that `CachedCMakePackage` already sets `CMAKE_PREFIX_PATH` as a `cmake` command line option (the behavior is inherited from `CMakePackage`): did you also need to set it in the cache?

bvanessen replied:
@scheibelp While it may set it as a command line option, if you run cmake outside of the spack environment that created it, it will not be able to find all of the spack created dependencies.  Preserving the prefix path inside of the `.cmake` file allows you to reuse the file afterwards and without spack being actively involved.

scheibelp replied:
OK: I assume the same principle applies to the generator (which is handled in `cmake.py` as a command-line option, but specifying it in the settings file avoids the need to set it when invoking `cmake` outside of Spack.

bvanessen replied:
@scheibelp That is correct. It will be cached in the `.cmake` file so that when you invoke it you have a command that looks like 
```
cmake -C<filename>.cmake -B <path>/build -DCMAKE_INSTALL_PREFIX=<path>/install .
```

benson31 replied:
This is actually not a path. [It's most appropriately a string in CMake](https://cmake.org/cmake/help/git-stage/command/set.html#set-cache-entry). A `PATH` is a path to a directory (note "a directory", not a list of directories).

scheibelp commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -357,181 +326,179 @@ def setup_build_environment(self, env):
+        sys_type = spec.architecture
+        if "SYS_TYPE" in env:
+            sys_type = env["SYS_TYPE"]
+        return sys_type
```
I see several `CachedCMakePackage`s use this: I assume they could be simplified accordingly.

scheibelp replied:
(to clarify: I'm not requesting that happens in this PR)

bvanessen replied:
@scheibelp I think that this is a convention, so I am not sure that it should be moved up to the `cached_cmake.py` package.

scheibelp commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -357,181 +326,179 @@ def setup_build_environment(self, env):
+    def cache_name(self):
+        hostname = socket.gethostname()
+        if "SYS_TYPE" in env:
+            hostname = hostname.rstrip("1234567890")
```
I'm guessing this is done for builds conducted from allocations (i.e. where the nodes get a number): I recommend making a comment about that (in particular note that it is entirely for cosmetic reasons).

benson31 replied:
All nodes on LC machines get numbers, even the login and/or launch nodes. This is just to get back to the canonical system name (e.g., `pascal` or `corona` or `lassen`).

scheibelp commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -357,181 +326,179 @@ def setup_build_environment(self, env):
+            entries.append(cmake_cache_string("CMAKE_SHARED_LINKER_FLAGS", "-fuse-ld=gold"))
+
+        # if "+rocm" in spec:
+        #     entries.insert(0, cmake_cache_path("CMAKE_CXX_COMPILER", spec["hip"].hipcc))
```
(question) do you want to keep this commented-out code? This should generally not be required

scheibelp commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -357,181 +326,179 @@ def setup_build_environment(self, env):
+            #         "-DHIP_ROOT_DIR={0}".format(spec["hip"].prefix),
+            #         "-DHIP_CXX_COMPILER={0}".format(self.spec["hip"].hipcc),
+            #     ]
+            # )
```
I think you can remove this as it is set in `CachedCMakePackage`

scheibelp commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -357,181 +326,179 @@ def setup_build_environment(self, env):
+        entries.append(
+            cmake_cache_string("CMAKE_MAKE_PROGRAM", "{0}/ninja".format(spec["ninja"].prefix.bin))
+        )
+        entries.append(cmake_cache_string("CMAKE_GENERATOR", "Ninja"))
```
Note that you already have `generator("ninja")` which should add this to the `cmake` command-line options by default (this should also occur for `CachedCMakePackage`s).

bvanessen replied:
@scheibelp I have pushed the definition of this into the cached_cmake.py package.

scheibelp commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -357,181 +326,179 @@ def setup_build_environment(self, env):
+        )
+        entries.append(self.define_cmake_cache_from_variant("LBANN_WITH_BOOST", "boost"))
+        entries.append(self.define_cmake_cache_from_variant("LBANN_WITH_CONDUIT", "conduit"))
+        entries.append(self.define_cmake_cache_from_variant("LBANN_WITH_NVSHMEM", "nvshmem"))
```
(minor) I suggest:

```
for opt, val in [("LBANN_WITH_CNPY", "numpy"), ("LBANN_DETERMINISTIC", "deterministic"), ...]:
  entries.append(self.define_cmake_cache_from_variant())
```

bvanessen replied:
Good suggestion.

scheibelp commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -316,39 +317,7 @@ class Lbann(CMakePackage, CudaPackage, ROCmPackage):
-            )
-
-        return args
+    depends_on("ninja", type="build")
```
This is inherited from `CMakePackage` (a superclass of `CachedCMakePackage`) so you should not have to define it here.

bvanessen replied:
Done.

scheibelp commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -63,6 +67,38 @@ def cache_name(self):
+        value = self.pkg.spec.variants[variant].value
+        if isinstance(value, (tuple, list)):
+            # Sort multi-valued variants for reproducibility
+            value = sorted(value)
```
We should already be doing this for all variants (regardless of build system) - are you seeing counterexamples to this?

bvanessen replied:
Removed.

scheibelp reviewed (changes requested):
At a high level:

* There are a couple potential issues
* There are a number of additions which should be redundant with what Spack/CMakePackage already do
* There are a couple spots where I think a comment would be useful

bvanessen committed:
Pushed the definition of the CMake generator into the
CachedCmakePackage.

Added legacy / deprecaed CUDA root toolkit env variable to support
BLT.

Made the substitution of the : for ; in CMAKE_PREFIX_PATH more robust.

Streamlined the mapping of variants to CMake variables for LBANN.

bvanessen committed:
Changed it so that FORCE is not the default option.

bvanessen committed:
[@spackbot] updating style on behalf of bvanessen

bvanessen committed:
Cleanup based on review feedback.
